### PR TITLE
[tlv] TLVReader::Get CharSpan to Stop on ASCII Unit Separator

### LINKED ISSUE #24419: [TLV] TLVReader::Get(CharSpan&) to Stop on ASCII Unit Separator
### Feature description

The TLVReader::Get(CharSpan&) should always stops at the first ASCII Unit Separator codepoint (ascii 0x1F).

The rationale is that clients should never see garbage at the end of string due to localization IDs.

### Platform

core

### Platform Version(s)

0f328000a33c02a98ef66e3e30f213c5c555a88e

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by emargolis**
Fixes #24419

The TLVReader::Get(CharSpan&) always stops at the first IS1 codepoint (ascii 0x1F).

The rationale is that clients should never see garbage at the end of string due to localization IDs.

### CODE REVIEW

File: src/lib/core/TLVReader.cpp [hunk 1]
```diff
@@ -300,14 +300,27 @@ CHIP_ERROR TLVReader::Get(ByteSpan & v)
...
         return CHIP_ERROR_WRONG_TLV_TYPE;
     }
 
     const uint8_t * bytes;
     ReturnErrorOnFailure(GetDataPtr(bytes)); // Does length sanity checks
-    v = CharSpan(Uint8::to_const_char(bytes), GetLength());
+
+    uint32_t len = GetLength();
+
+    // If ASCII Unit Separator (0x1f) is present in the string then method returns string ending
```
Comment (tcarmelveilleux) [inline src/lib/core/TLVReader.cpp:315]: Let's call it Information Separator 1 to match Unicode and spec text

File: src/lib/core/tests/TestTLV.cpp [hunk 2]
```diff
@@ -2789,6 +2789,52 @@ void CheckTLVByteSpan(nlTestSuite * inSuite, void * inContext)
...
+        const char * expectedString;
+    };
+
+    // clang-format off
+    static CharSpanTestCase sCharSpanTestCases[] = {
+        // Test String                                                   Expected String from Get()
+        // =========================================================================================
+        {  "This is a test case #0",                                     "This is a test case #0"  },
+        {  "This is a test case #1\x1fTest Localized String Identifier", "This is a test case #1"  },
+        {  "This is a test case #2 \x1f abc \x1f def",                   "This is a test case #2 " },
```
Comment (tcarmelveilleux) [inline src/lib/core/tests/TestTLV.cpp:2806]: Please add a case that ends with IS1 alone as well

**Review by tcarmelveilleux — APPROVED** [review]

File: src/lib/core/TLVReader.cpp (hunk 1 above)
Comment (emargolis) [reply src/lib/core/TLVReader.cpp]: done

File: src/lib/core/tests/TestTLV.cpp (hunk 2 above)
Comment (emargolis) [reply src/lib/core/tests/TestTLV.cpp:2806]: done

File: src/lib/core/TLVReader.cpp [hunk 3]
```diff
@@ -300,14 +300,27 @@ CHIP_ERROR TLVReader::Get(ByteSpan & v)
...
 
     const uint8_t * bytes;
     ReturnErrorOnFailure(GetDataPtr(bytes)); // Does length sanity checks
-    v = CharSpan(Uint8::to_const_char(bytes), GetLength());
+
+    uint32_t len = GetLength();
+
+    // If Unicode Information Separator 1 (0x1f) is present in the string then method returns
+    // string ending at first appearance of the Information Separator 1.
+    const uint8_t * unitSeparator = reinterpret_cast<const uint8_t *>(memchr(bytes, kUnicodeInformationSeparator1, len));
```
Comment (bzbarsky-apple) [inline src/lib/core/TLVReader.cpp:317]:
```suggestion
    const uint8_t * infoSeparator = reinterpret_cast<const uint8_t *>(memchr(bytes, kUnicodeInformationSeparator1, len));
```
Perhaps?

**Review by bzbarsky-apple — APPROVED** [review]

File: src/lib/core/TLVReader.cpp (hunk 3 above)
Comment (emargolis) [reply src/lib/core/TLVReader.cpp]: fixed

**Pull request merged**